packages/attestation-service/src/attestation.ts

```diff
@@ -1,38 +1,128 @@
+
+export const AttestationRequestType = t.type({
+  phoneNumber: E164PhoneNumberType,
+  account: AddressType,
+  issuer: AddressType,
+})
+
+export type AttestationRequest = t.TypeOf<typeof AttestationRequestType>
+
+function getAttestationKey() {
   if (process.env.ATTESTATION_KEY === undefined) {
     console.error('Did not specify ATTESTATION_KEY')
     throw new Error('Did not specify ATTESTATION_KEY')
   }
 
-  const signature = attestToIdentifier(phoneNumber, account, process.env.ATTESTATION_KEY)
+  return process.env.ATTESTATION_KEY
+}
+
+async function validateAttestationRequest(request: AttestationRequest) {
+  // check if it exists in the database
+  if (
+    (await existingAttestationRequest(request.phoneNumber, request.account, request.issuer)) !==
+    null
+  ) {
+    throw new Error('Attestation already sent')
+  }
+  const key = getAttestationKey()
+  const address = privateKeyToAddress(key)
+
+  // TODO: Check with the new Accounts.sol
+  if (address.toLowerCase() !== request.issuer.toLowerCase()) {
+    throw new Error(`Mismatching issuer, I am ${address}`)
+  }
+
+  const attestations = await kit.contracts.getAttestations()
+  const state = await attestations.getAttestationState(
+    request.phoneNumber,
+    request.account,
+    request.issuer
+  )
+
+  console.info(state)
+  if (state.attestationState !== AttestationState.Incomplete) {
+    throw new Error('No incomplete attestation found')
+  }
+
+  // TODO: Check expiration
+  return
+}
+
+async function validateAttestation(
+  attestationRequest: AttestationRequest,
+  attestationCode: string
+) {
+  const attestations = await kit.contracts.getAttestations()
+  const isValid = await attestations.validateAttestationCode(
+    attestationRequest.phoneNumber,
+    attestationRequest.account,
+    attestationRequest.issuer,
+    attestationCode
+  )
+  if (!isValid) {
+    throw new Error('Valid attestation could not be provided')
+  }
+  return
+}
+
+function signAttestation(phoneNumber: E164Number, account: Address) {
+  const signature = attestToIdentifier(phoneNumber, account, getAttestationKey())
 
   return SignatureUtils.serializeSignature(signature)
 }
 
 function toBase64(str: string) {
-  return Buffer.from(str, 'hex').toString('base64')
+  return Buffer.from(str.slice(2), 'hex').toString('base64')
 }
 
 function createAttestationTextMessage(attestationCode: string) {
   return `<#> ${toBase64(attestationCode)} ${process.env.APP_SIGNATURE}`
 }
 
-export async function handleAttestationRequest(req: express.Request, res: express.Response) {
-  // TODO: Should parse request appropriately
-
-  // TODO: Should validate request here
-  // const attestations = await kit.contracts.getAttestations()
-
-  // Produce attestation
-  const attestationCode = signAttestation(req.body.phoneNumber, req.body.account)
-  const textMessage = createAttestationTextMessage(attestationCode)
+export async function handleAttestationRequest(
+  _req: express.Request,
```

Comment:
why the underscore?

Reply:
In typescript (and many other languages), a leading underscore indicates the parameter not being used. Not prefixing is I believe a lint violation (or maybe even compiler?)

Reply:
huh, didn't know that!